src/app/util/util.service.ts

```diff
@@ -53,4 +62,104 @@ export class UtilService {
+  // At the moment, the LaTeX source code is persisted in the backend
+
+  // public getBlobFromDataUri(dataURI: string): Blob {
+  //   // convert base64 to raw binary data held in a string
+  //   // doesn't handle URLEncoded DataURIs - see SO answer #6850276 for code that does this
+  //   const byteString = atob(dataURI.split(',')[1]);
+  //
+  //   // separate out the mime component
+  //   const mimeString = dataURI.split(',')[0].split(':')[1].split(';')[0];
+  //
+  //   // write the bytes of the string to an ArrayBuffer
+  //   const ab = new ArrayBuffer(byteString.length);
+  //   const ia = new Uint8Array(ab);
+  //   for (let i = 0; i < byteString.length; i++) {
+  //     ia[i] = byteString.charCodeAt(i);
+  //   }
+  //
+  //   // write the ArrayBuffer to a blob
+  //   return new Blob([ab], { type: mimeString });
+  // }
+  //
+  // public getDataUriFromBlob(blob: Blob): string {
+  //   const fileReader = new FileReader();
+  //   fileReader.readAsDataURL(blob);
+  //   let dataUri: string;
+  //   fileReader.onload = () => {
+  //     dataUri = fileReader.result.toString();
+  //   };
+  //   return dataUri;
+  // }
+
+  public packTextAndPackages(text = '', packages: string[]): string {
+    return text.concat(packages.join(''));
+  }
+
+  unpackTextAndPackages(
+    packedData: string
+  ): { latexContent: string; latexPackages: string[] } {
+    const splitData = packedData.split('\\use');
+    const content = splitData[0];
+    const packages: string[] = [];
+    for (let i = 1; i < splitData.length; i++) {
+      packages.push('\\use' + splitData[i]);
+    }
+    return { latexContent: content, latexPackages: packages };
+  }
+
+  public renderPackedDataAndReturnUrlToPdfBlob(
+    packedData: string,
+    output = 'pdf'
+  ): Observable<string> {
+    const data = this.unpackTextAndPackages(packedData);
+    return this.renderLatexContentAndReturnUrlToPdfBlob(
+      data.latexContent,
+      data.latexPackages,
+      output
+    );
+  }
+
+  public renderLatexContentAndReturnUrlToPdfBlob(
+    latexContent: string,
+    additionalPackages: string[] = [],
+    output = 'pdf'
+  ): Observable<string> {
+    const packages = this.getDefaultLatexPackages();
+    if (additionalPackages) {
+      for (const additionalPackage of additionalPackages) {
+        if (!packages.includes(additionalPackage)) {
+          packages.push(additionalPackage);
+        }
+      }
+    }
+    const latexBody: LatexContent = {
+      content: latexContent,
+      latexPackages: packages,
+      output,
+    };
+    return this.latexRendererService.renderLatexAsPdf({ body: latexBody }).pipe(
+      map((response: string[]) => {
+        if (response) {
+          const latexBlob = this.createBlobFromRenderedResult(response);
+          return URL.createObjectURL(latexBlob);
+        }
+      })
+    );
+  }
+
+  public getDefaultLatexPackages(): string[] {
+    return ['\\usepackage{tikz}', '\\usetikzlibrary{quantikz}'];
```

Comment:
Can we place this somewehre, e.g. in a sperate constants file (maybe `latex-service-constants.ts`), to easily locate is. I presume searching for this could be time consuming.

Assuming one does not use a powerful IDE :)

Reply:
done